### PR TITLE
Bidirectional Dollar Cost Average Bridge 

### PR DESCRIPTION
**Pull request by LHerskind**
# Description

An implementation of a DeFi bridge that let users open a DCA position, striving to swap x A to y B over z time. Time is defined in `ticks` to allow aggregation of multiple DCA positions. 

The bridge allows both positions going A -> B and B -> A, giving us the ability to match these directly before needing to sell tokens to external parties. e.g., if Alice want to swap A -> B and bob B -> A, for a given tick, one of them can be filled using the others DCA.

To handle the excess (stuff we cannot match between DCAs directly), we can use an external party, who might buy without slippage using some oracle price. And if no-one is interested in that, the excess should be traded directly through a dex etc.

# Checklist:

- [ ] I have reviewed my diff in github, line by line.
- [ ] Every change is related to the PR description.
- [ ] There are no unexpected formatting changes, or superfluous debug logs.
- [ ] The branch has been rebased against the head of its merge target.
- [ ] I'm happy for the PR to be merged at the reviewers next convenience.

### CODE REVIEW

**Review by benesjan — CHANGES_REQUESTED** [review]
Currently only pointed out naming issues. Can you please finish full NatSpec before I review it again? The bridge is complex and I spend too much time decoding what the params and global vars mean.

File: src/bridges/dca/BiDCABridge.sol [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+        pokeTicks(nextTick, _ticks);
+    }
+
+    /**
+     * @notice Helper used to poke storage of ticks to make deposits more consistent in gas usage
+     * @dev First sstore is very expensive, so by doing it as this, we can prepare it before deposit
+     * @param _startTick The first tick to poke
+     * @param _ticks The number of ticks to poke
+     */
+    function pokeTicks(uint256 _startTick, uint256 _ticks) public {
```
[the remark this answers is not in the document]
Comment (LHerskind) [reply src/bridges/dca/BiDCABridge.sol:131]: `pokeNextTicks` are calling this function

File: src/bridges/dca/BiDCABridge.sol [hunk 2]
```diff
@@ -0,0 +1,648 @@
...
+import {SafeCastLib} from "./SafeCastLib.sol";
+
+/**
+ * @notice Initial abstract implementation of "Dollar" Cost Averaging.
+ * Supports bidirectional "selling", e.g., A -> B and B -> A in same bridge
+ * Will match internal orders first, before allowing external parties to trade.
+ * @dev Built for assets with 18 decimals precision
+ * @dev A contract that inherits must handle the case for forcing a swap through a DEX.
+ * @author Lasse Herskind (LHerskind on GitHub).
+ */
```
Comment (benesjan) [inline src/bridges/dca/BiDCABridge.sol:24]: Can you please describe here how the bridge internally works? How the ticks are indexed by time, how does rebalancing work etc.

File: src/bridges/dca/BiDCABridge.sol [hunk 3]
```diff
@@ -16,8 +16,18 @@ import {SafeCastLib} from "./SafeCastLib.sol";
...
+ * The bridge implements a bidirectional dollar cost averaging protocol,
+ * allowing users to go either direction between two tokens A and B.
+ * The "order" is executed over a period of time, instead of all at once.
+ * For Eth and Dai, this allows the user to sell Dai to buy Eth over X days, and vice versa.
+ * The timeperiod is divided into ticks, with each tick keeping track of the funds that should be traded in that tick.
+ * As well as the price, and how much have been received in return.
+ * To "balance" the scheme, an external party can buy assets at the orace-price (no slippage).
+ * The amount that can be bought by the external party, depends on the ticks and how much can be matched internally.
+ * As part of the balancing act, each tick will match the A and B holdings it has to sell, using the oracle price
+ * (or infer price from prior price and current price). The excess from this internal matching is then sold off (of as much as possible).
```
Comment (benesjan) [inline src/bridges/dca/BiDCABridge.sol:28]: "(of as much as possible)" sounds a bit weird. Should it be "(for as much as possible)"?

File: src/bridges/dca/UniswapDCABridge.sol [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+/**
+ * @notice Initial implementation of the BiDCA bridge using Uniswap as the DEX and Chainlink as oracle.
+ * The bridge is using DAI and WETH for A and B with a Chainlink oracle to get prices.
+ * The bridge allows users to force a rebalance through Uniswap pools.
+ * The forced rebalance will:
+ * 1. rebalance internally in the ticks, as the BiDCABridge,
+ * 2. rebalance cross-ticks, e.g., excess from the individual ticks are matched,
+ * 3. swap any remaining excess using Uniswap, and rebalance with the returned assets
+ * @dev The `MAX_AGE` of a price feed, means that a abandoned price-feed will brick the bridge.
```
Comment (benesjan) [inline src/bridges/dca/UniswapDCABridge.sol:33]:
```suggestion
 * @dev The `MAX_AGE` of a price feed, means that an abandoned price-feed will brick the bridge.
```

File: src/specs/dca/readme.md [hunk 5]
```diff
@@ -51,6 +51,7 @@ When using the chainlink oracle, this is centralization issue as chainlink could
 A position needs to be finalised before it can be exited.
 For a position to be ready to be finalised, all its available funds must have been sold for the opposite asset.
 When finalised, the accumulated funds will be returned to the bridge and the user may claim them on L2.
+To incentivise searchers on L1 to finalise the transactions, a fee of 10 basis points of the swapped assets are paid to the initiator of the transaction (`tx.origin` instead of `msg.sender` as the rollup will always be the `msg.sender` and we cannot pass any address along).
```
Comment (benesjan) [inline src/specs/dca/readme.md:54]: Fee is specified in constructor so should be saying that instead of fixed 10 basis point fee.

**Review by benesjan — COMMENTED** [review]
Changes look good to me. The comments clarified it a lot and the incentive is a good idea.

**Pull request merged**